kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/HighOffsetMetricsExposer.scala

```diff
@@ -0,0 +1,121 @@
+              val increase = partitionInfo.endOffset - lastValue
+              metricUpdates = metricUpdates :+ (aggregateCounter, increase)
+              if (config.perPartitionMetrics) {
+                metricUpdates = metricUpdates :+ (counter, increase)
+              }
+              Some((counter, partitionInfo.endOffset))
+            case None =>
+              val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
+              metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
+              if (config.perPartitionMetrics) {
+                metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
+              }
+              Some((counter, partitionInfo.endOffset))
+          }
+          Some((newTopicMap, aggregateCounter))
+        case None =>
+          val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
+          val newTopicMap = Map.newBuilder
+            .addOne(
+              topicPartition.partition,
+              (counter, partitionInfo.endOffset)
+            )
+            .result()
+          val aggregateCounter = Helper.newTopicMetric(topicPartition.topic)
+          metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
+          if (config.perPartitionMetrics) {
+            metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
+          }
+          Some((newTopicMap, aggregateCounter))
+      }
+    }
+
+    // Update the metric map to reflect removals
+    for (topicPartition <- removed) {
+      val topic = topicPartition.topic
+      val partition = topicPartition.partition
+
+      newMetricMap = newMetricMap.updatedWith(topic) {
+        case Some((map: Map[Int, (Metric.Counter[Long], Long)], aggregateCounter: Metric.Counter[Long])) =>
+          val oldMetricValue = map.get(partition).map(_._2).getOrElse(0L)
+          val newMap = map.removed(partition)
+          metricUpdates = metricUpdates :+ (aggregateCounter, -oldMetricValue)
+          Some((newMap, aggregateCounter))
+        case None => None
+      }
+    }
+
+    // Clean up topics with no partitions
+    newMetricMap = newMetricMap.filter((_, v) => v._1.nonEmpty)
+
+    (MetricMap(newMetricMap), metricUpdates)
+  }
+}
+
+object HighOffsetMetricsExposer {
+  def startPublishing(
+      topicInfos: ZStream[MetricExposerConfig, Nothing, KafkaTopicPartitionInfoChanges]
+  ): URIO[MetricExposerConfig, Fiber.Runtime[Nothing, Unit]] =
+    for {
+      config <- ZIO.service[MetricExposerConfig]
+      fibre <- topicInfos
+        .runFoldZIO(MetricMap(Map.empty)) { (metricMap: MetricMap, topicPartitionInfoChanges: KafkaTopicPartitionInfoChanges) =>
+          val (newMetricMap, metricUpdates) = metricMap.updateWithTopicParitionChanges(topicPartitionInfoChanges, config)
+          ZIO.foreachDiscard(metricUpdates)((metric, increase) => metric.update(increase)).as(newMetricMap)
```

Comment:
`metric.update(increase)` could be `metric.incrementBy(increase)` to be more readable?